From df742928bee77f31bd1c3c845720a7e3fa06c10b Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Thu, 18 Apr 2024 09:54:11 +0200
Subject: [PATCH 0789/1695] clk: stm32mp21: add clock dependency from BSEC DENR

The CoreSight devices can be accesses only when BSEC_DENR:DBGSWEN
is set, otherwise a bus-error occurs.
To prevent Linux from accessing unavailable CoreSight devices,
make the exported SCMI clock of the CoreSight bus clock depending
from BSEC_DENR:DBGSWEN.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git e7d1da9be1ae927499110fa8e756da92a14c6d06 clk: stm32mp21: add clock dependency from BSEC DENR]
Signed-off-by: Antonio Borneo <antonio.borneo@foss.st.com>
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Change-Id: I5992c4f12c0654fb0eefb27ec14b1b2ab2653706
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/374581
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/clk/clk-stm32mp21.c              | 56 ++++++++++++++++---
 .../dt-bindings/clock/st,stm32mp21-rcc.h      | 10 +++-
 2 files changed, 55 insertions(+), 11 deletions(-)

diff --git a/core/drivers/clk/clk-stm32mp21.c b/core/drivers/clk/clk-stm32mp21.c
index 9c54df023..27b327c8f 100644
--- a/core/drivers/clk/clk-stm32mp21.c
+++ b/core/drivers/clk/clk-stm32mp21.c
@@ -8,6 +8,7 @@
 #include <libfdt.h>
 #include <drivers/clk.h>
 #include <drivers/clk_dt.h>
+#include <drivers/stm32_bsec.h>
 #include <drivers/stm32_rif.h>
 #include <drivers/stm32mp21_rcc.h>
 #include <dt-bindings/clock/st,stm32mp21-rcc.h>
@@ -2881,6 +2882,21 @@ static const struct clk_ops clk_stm32_rif_gate_ops = {
 	.restore_context = clk_stm32_rif_gate_pm_restore,
 };
 
+static TEE_Result clk_stm32_rif_gate_dbg_enable(struct clk *clk)
+{
+	if (!stm32_bsec_self_hosted_debug_is_enabled())
+		return TEE_ERROR_ACCESS_DENIED;
+
+	return clk_stm32_rif_gate_enable(clk);
+}
+
+static const struct clk_ops  clk_stm32_rif_gate_dbg_ops = {
+	.enable		= clk_stm32_rif_gate_dbg_enable,
+	.disable	= clk_stm32_rif_gate_disable,
+	.is_enabled	= clk_stm32_rif_gate_is_enabled,
+	.restore_context = clk_stm32_rif_gate_pm_restore,
+};
+
 struct clk_stm32_rif_composite_cfg {
 	int sec_id;
 	int gate_id;
@@ -3105,6 +3121,19 @@ static const struct clk_ops ck_timer_ops = {
 		.parents = { (_parent) },\
 	}
 
+#define RIF_GATE_DBG(_name, _parent, _flags, _gate_id, _sec_id)\
+	struct clk _name = {\
+		.ops = &clk_stm32_rif_gate_dbg_ops,\
+		.priv = &(struct clk_stm32_rif_gate_cfg) {\
+			.sec_id	= (_sec_id),\
+			.gate_id = (_gate_id),\
+		},\
+		.name = #_name,\
+		.flags = (_flags),\
+		.num_parents = 1,\
+		.parents = { (_parent) },\
+	}
+
 #define RIF_COMPOSITE(_name, _nb_parents, _parents, _flags,\
 			_gate_id, _div_id, _mux_id, _sec_id)\
 	struct clk _name = {\
@@ -3243,7 +3272,8 @@ static STM32_DIVIDER(ck_icn_apb2, &ck_icn_ls_mcu, 0, DIV_APB2);
 static STM32_DIVIDER(ck_icn_apb3, &ck_icn_ls_mcu, 0, DIV_APB3);
 static STM32_DIVIDER(ck_icn_apb4, &ck_icn_ls_mcu, 0, DIV_APB4);
 static STM32_DIVIDER(ck_icn_apb5, &ck_icn_ls_mcu, 0, DIV_APB5);
-static STM32_DIVIDER(ck_icn_apbdbg, &ck_icn_ls_mcu, 0, DIV_APBDBG);
+static RIF_COMPOSITE(ck_icn_apbdbg, 1, { &ck_icn_ls_mcu }, 0,
+		     GATE_DBG, DIV_APBDBG, NO_MUX, RCC_RIF_DEBUG_TRACE);
 
 #define STM32_TIMER(_name, _parent, _flags, _apbdiv, _timpre)\
 	struct clk _name = {\
@@ -3267,14 +3297,22 @@ static RIF_GATE(ck_icn_p_dbgmcu, &ck_icn_apb3, 0,
 		GATE_DBGMCU, RCC_RIF_DEBUG_TRACE);
 static RIF_GATE(ck_dap, &ck_hsi, 0, GATE_DBGMCU, RCC_RIF_DEBUG_TRACE);
 
-static RIF_GATE(ck_sys_dbg, &ck_icn_apbdbg, 0, GATE_DBG, RCC_RIF_DEBUG_TRACE);
-static RIF_GATE(ck_icn_s_stm, &ck_icn_ls_mcu, 0, GATE_STM,
-		RCC_RIF_DEBUG_TRACE);
+static RIF_GATE_DBG(ck_sys_dbg, &ck_icn_apbdbg, 0, GATE_DBG,
+		    RCC_RIF_DEBUG_TRACE);
+static RIF_GATE_DBG(ck_icn_p_stm, &ck_icn_apbdbg, 0, GATE_STM,
+		    RCC_RIF_DEBUG_TRACE);
+static RIF_GATE_DBG(ck_icn_s_stm, &ck_icn_ls_mcu, 0, GATE_STM,
+		    RCC_RIF_DEBUG_TRACE);
 static RIF_GATE(ck_ker_tsdbg, &ck_flexgen_43, 0, GATE_DBG, RCC_RIF_DEBUG_TRACE);
 static RIF_GATE(ck_ker_tpiu, &ck_flexgen_44, 0, GATE_TRACE,
 		RCC_RIF_DEBUG_TRACE);
-static RIF_GATE(ck_icn_m_etr, &ck_flexgen_45, 0, GATE_ETR, RCC_RIF_DEBUG_TRACE);
-static RIF_GATE(ck_sys_atb, &ck_flexgen_45, 0, GATE_DBG, RCC_RIF_DEBUG_TRACE);
+static RIF_GATE_DBG(ck_icn_p_etr, &ck_icn_apbdbg, 0, GATE_ETR,
+		    RCC_RIF_DEBUG_TRACE);
+static RIF_GATE_DBG(ck_icn_m_etr, &ck_flexgen_45, 0, GATE_ETR,
+		    RCC_RIF_DEBUG_TRACE);
+static RIF_GATE_DBG(ck_sys_atb, &ck_flexgen_45, 0, GATE_DBG,
+		    RCC_RIF_DEBUG_TRACE);
+
 static RIF_GATE(ck_icn_s_sysram, &ck_icn_hs_mcu, 0, GATE_SYSRAM,
 		RCC_RIF_SYSRAM);
 static RIF_GATE(ck_icn_s_retram, &ck_icn_hs_mcu, 0, GATE_RETRAM,
@@ -3644,7 +3682,8 @@ static struct clk *stm32mp21_clk_provided[STM32MP21_ALL_CLK_NB] = {
 	[CK_BUS_GPIOI]		= &ck_icn_p_gpioi,
 	[CK_BUS_GPIOZ]		= &ck_icn_p_gpioz,
 	[CK_BUS_RTC]		= &ck_icn_p_rtc,
-	[CK_BUS_STM]		= &ck_icn_s_stm,
+	[CK_BUS_STM]		= &ck_icn_p_stm,
+	[CK_KER_STM]		= &ck_icn_s_stm,
 	[CK_BUS_FMC]		= &ck_icn_p_fmc,
 	[CK_BUS_ETH1]		= &ck_icn_p_eth1,
 	[CK_BUS_ETH2]		= &ck_icn_p_eth2,
@@ -3780,7 +3819,8 @@ static struct clk *stm32mp21_clk_provided[STM32MP21_ALL_CLK_NB] = {
 	[CK_KER_LPTIM5]		= &ck_ker_lptim5,
 	[CK_KER_TSDBG]		= &ck_ker_tsdbg,
 	[CK_KER_TPIU]		= &ck_ker_tpiu,
-	[CK_BUS_ETR]		= &ck_icn_m_etr,
+	[CK_BUS_ETR]		= &ck_icn_p_etr,
+	[CK_KER_ETR]		= &ck_icn_m_etr,
 	[CK_BUS_SYSATB]		= &ck_sys_atb,
 	[CK_KER_OSPI1]		= &ck_ker_ospi1,
 	[CK_KER_FMC]		= &ck_ker_fmc,
diff --git a/core/include/dt-bindings/clock/st,stm32mp21-rcc.h b/core/include/dt-bindings/clock/st,stm32mp21-rcc.h
index b72ea7845..1b6d533e1 100644
--- a/core/include/dt-bindings/clock/st,stm32mp21-rcc.h
+++ b/core/include/dt-bindings/clock/st,stm32mp21-rcc.h
@@ -308,8 +308,10 @@
 #define HSE_DIV2_CK		285
 #define CK_DBGMCU		286
 #define CK_DAP			287
+#define CK_KER_ETR		288
+#define CK_KER_STM		289
 
-#define STM32MP21_LAST_CLK	288
+#define STM32MP21_LAST_CLK	290
 
 #define CK_SCMI_ICN_HS_MCU	0
 #define CK_SCMI_ICN_SDMMC	1
@@ -397,7 +399,7 @@
 #define CK_SCMI_TIMG2		83
 #define CK_SCMI_BKPSRAM		84
 #define CK_SCMI_BSEC		85
-#define CK_SCMI_ETR		86
+#define CK_SCMI_BUS_ETR		86
 #define CK_SCMI_FMC		87
 #define CK_SCMI_GPIOA		88
 #define CK_SCMI_GPIOB		89
@@ -421,6 +423,8 @@
 #define CK_SCMI_SYSDBG		107
 #define CK_SCMI_SYSATB		108
 #define CK_SCMI_TSDBG		109
-#define CK_SCMI_STM		110
+#define CK_SCMI_BUS_STM		110
+#define CK_SCMI_KER_STM		111
+#define CK_SCMI_KER_ETR		112
 
 #endif /* _DT_BINDINGS_STM32MP21_CLKS_H_ */
-- 
2.47.2

